### PR TITLE
Update SHUTDOWN and Signal Handling

### PR DESCRIPTION
**Pull request by zuiderkwast**
Doc update for redis/redis#9872

### CODE REVIEW

File: commands/info.md [hunk 1]
```diff
@@ -71,6 +71,9 @@ Here is the meaning of all fields in the **server** section:
 *   `lru_clock`: Clock incrementing every minute, for LRU management
 *   `executable`: The path to the server's executable
 *   `config_file`: The path to the config file
+*   `io_threads_active`: Flag indicating if I/O threads are active
+*   `shutdown_in_seconds`: The maximum time remaining for replicas to catch up the replication before shutting completing the shutdown sequence.
```
Comment (yoav-steinberg) [inline commands/info.md:75]:
```suggestion
*   `shutdown_in_seconds`: The maximum time remaining for replicas to catch up the replication before completing the shutdown sequence.
```

File: commands/shutdown.md [hunk 2]
```diff
@@ -1,35 +1,40 @@
...
+* If there are any replicas lagging behind in replication:
+  * Pause clients attempting to write by performing a `CLIENT PAUSE` with the `WRITE` option.
+  * Wait up to the configured `shutdown-timeout` (default 10 seconds) for replicas to catch up the replication offset.
 * Stop all the clients.
 * Perform a blocking SAVE if at least one **save point** is configured.
 * Flush the Append Only File if AOF is enabled.
 * Quit the server.
 
 If persistence is enabled this commands makes sure that Redis is switched off
 without the lost of any data.
```
Comment (yoav-steinberg) [inline commands/shutdown.md:12]:
```suggestion
without any data loss.
```

File: commands/shutdown.md [hunk 3]
```diff
@@ -1,35 +1,40 @@
...
-This is not guaranteed if the client uses simply `SAVE` and then `QUIT` because
-other clients may alter the DB data between the two commands.
```
Comment (yoav-steinberg) [inline commands/shutdown.md:11]: Why did you remove this comment? Because it goes without saying and isn't related to `SHUTDOWN` or for some other reason?

File: commands/shutdown.md [hunk 4]
```diff
@@ -1,35 +1,40 @@
...
-* **SHUTDOWN SAVE** will force a DB saving operation even if no save points are
-  configured.
-* **SHUTDOWN NOSAVE** will prevent a DB saving operation even if one or more
-  save points are configured.
-  (You can think of this variant as an hypothetical **ABORT** command that just
-  stops the server).
+* **SAVE** will force a DB saving operation even if no save points are configured.
+* **NOSAVE** will prevent a DB saving operation even if one or more save points are configured.
+* **NOW** skips waiting for lagging replicas, i.e. it bypasses the first step in the shutdown sequence.
+* **FORCE** ignores any errors that would normally prevent the server from exiting.
```
Comment (yoav-steinberg) [inline commands/shutdown.md:30]: Not clear that such "any errors" can be? Maybe add "details below" or reference the next section?

File: commands/shutdown.md [hunk 5]
```diff
@@ -41,11 +46,28 @@ instead. This happens when:
```
Comment (yoav-steinberg) [inline commands/shutdown.md:46]: See above:
> there are two conditions

Should now be "three conditions".

File: commands/shutdown.md [hunk 6]
```diff
@@ -41,11 +46,28 @@ instead. This happens when:
...
+The first command will turn off the AOF if needed, and will terminate the AOF rewriting child if there is one active.
+The second command will not have any problem to execute since the AOF is no longer enabled.
 
-There are conditions when we want just to terminate a Redis instance ASAP, regardless of what its content is. In such a case, the right combination of commands is to send a **CONFIG appendonly no** followed by a **SHUTDOWN NOSAVE**. The first command will turn off the AOF if needed, and will terminate the AOF rewriting child if there is one active. The second command will not have any problem to execute since the AOF is no longer enabled.
+## Minimize the risk of data loss
+
+Since Redis 7.0, the server waits for lagging replicas up to a configurable `shutdown-timeout`, by default 10 seconds, before shutting down.
+This provides a best effort minimizing the risk of data loss in a situation where no save points are configured and AOF is disabled.
+Before version 7.0, shutting down a heavily loaded master node in a diskless setup is more likely to result in data loss.
```
Comment (yoav-steinberg) [inline commands/shutdown.md:61]:
```suggestion
Before version 7.0, shutting down a heavily loaded master node in a diskless setup was more likely to result in data loss.
```

File: topics/signals.md [hunk 7]
```diff
@@ -22,14 +22,24 @@ executed just after the script is killed, or if terminates spontaneously.
...
 * The server exits with an exit code of zero.
 
-In case the RDB file can't be saved, the shutdown fails, and the server continues to run in order to ensure no data loss. Since Redis 2.6.11 no further attempt to shut down will be made unless a new `SIGTERM` will be received or the `SHUTDOWN` command issued.
+In case the RDB file can't be saved, the shutdown fails, and the server continues to run in order to ensure no data loss.
+Likewise, if the user just turned on AOF, and the server triggered the first AOF rewrite in order to create the initial AOF file and this file can't be saved, the shutdown fails and the server continues to run.
+Since Redis 2.6.11 no further attempt to shut down will be made unless a new `SIGTERM` is received or the `SHUTDOWN` command issued.
+
+Since Redis 7.0, the server waits for lagging replicas up to a configurable `shutdown-timeout`, by default 10 seconds, before shutting down.
+This provides a best effort minimizing the risk of data loss in a situation where no save points are configured and AOF is disabled.
+Before version 7.0, shutting down a heavily loaded master node in a diskless setup is more likely to result in data loss.
```
Comment (yoav-steinberg) [inline topics/signals.md:41]:
```suggestion
Before version 7.0, shutting down a heavily loaded master node in a diskless setup was more likely to result in data loss.
```

File: commands/shutdown.md (hunk 3 above)
Comment (zuiderkwast) [reply commands/shutdown.md:11]: I think the comment was confusing and irrelevevant. QUIT is just closing the client's connection. It's not shutting down the server.

File: commands.json [hunk 8]
```diff
@@ -3912,6 +3912,22 @@
           "SAVE"
         ],
         "optional": true
+      },
+      {
+        "name": "now",
```
Comment (zuiderkwast) [inline commands.json:3917]: IIUC, this file is generated from redis and I should not touch this file. I'll revert it. Is that right?

File: commands/shutdown.md (hunk 6 above)
Comment (zuiderkwast) [reply commands/shutdown.md:61]: Right. I was thinking about users who continue to run old versions. For them it "is" still the case, but if we use "Before" we should also use "was". Another way to describe it would be "In versions prior to 7.0, ... is ....", but I like "before" because it's simpler language.

File: commands.json (hunk 8 above)
Comment (yoav-steinberg) [reply commands.json:3917]: As long as we're still using this version of the docs/redis-io (which doesn't use the generated file from the redis repo) we should keep both the redis repo and the redis-doc repo manually in sycn. So don't revert.

**Review by yoav-steinberg — APPROVED** [review]

File: commands/shutdown.md (hunk 5 above)
Comment (itamarhaber) [reply commands/shutdown.md:47]: "several"?

File: commands/shutdown.md [hunk 9]
```diff
@@ -1,51 +1,74 @@
...
 
 @return
 
 @simple-string-reply on error.
 On success nothing is returned since the server quits and the connection is
 closed.
+
+@history
+
+* `>= 7.0`: NOW and FORCE modifiers added.
```
Comment (itamarhaber) [inline commands/shutdown.md:73]:
```suggestion
* `>= 7.0.0`: `NOW` and `FORCE `modifiers added.
```

**Review by itamarhaber — APPROVED** [review]

File: commands/shutdown.md (hunk 5 above)
Comment (zuiderkwast) [reply commands/shutdown.md:47]: "all of the conditions below"... Many possibilitites

Comment (zuiderkwast) [reply commands/shutdown.md:47]: Looking at this again, it doesn't seem like conditions that conditions have to be fulfilled. It's rather two different *situations* where exit is prevented. Maybe it should instead be sth like

"However there are two situations where it is unsafe to do so, and unless the **FORCE** modifier is specified, the **SHUTDOWN** command will be refused with an error instead. This happens when:"

WDYT?

File: commands/shutdown.md [hunk 10]
```diff
@@ -70,5 +70,5 @@ closed.
 
 @history
 
-* `>= 7.0`: NOW and FORCE modifiers added.
+* `>= 7.0.0`: `NOW` and `FORCE `modifiers added.
```
Comment (yoav-steinberg) [inline commands/shutdown.md:73]:
```suggestion
* `>= 7.0.0`: `NOW` and `FORCE` modifiers added.
```

File: commands/shutdown.md (hunk 5 above)
Comment (zuiderkwast) [reply commands/shutdown.md:47]: See commit "conditions -> situations"

File: commands/shutdown.md [hunk 11]
```diff
@@ -1,51 +1,74 @@
...
-On success nothing is returned since the server quits and the connection is
-closed.
+@simple-string-reply: `OK` if `ABORT` was specified and shutdown was aborted.
+On successful shutdown, nothing is returned since the server quits and the connection is closed.
+On failure, an error is returned.
+
+@history
+
+* `>= 7.0.0`: `NOW`, `FORCE` and `ABORT` modifiers added.
+  Waiting for lagging replicas before exiting added.
```
Comment (itamarhaber) [inline commands/shutdown.md:74]: History is now recorded in the Redis repo, in each command's src/commands JSON file. Please move this tidbit there.

Comment (zuiderkwast) [reply commands/shutdown.md:74]: ACK. I'll need to add it in the Redis PR then.

**Review by itamarhaber — APPROVED** [review]
